Feature/srv-270

nest1ing opened the pull request.

nest1ing committed:
Add software table like in REST/DBus

Firmware images will placed into table
```
snmptranslate -Tp -IR YADRO-MIB::yadro
+--yadro(49769)
   |
...
   |
   +--yadroSoftwareTable(3)
      |
      +--yadroSoftwareEntry(1)
         |  Index: yadroSoftwareHash
         |
         +-- ---- String    yadroSoftwareHash(1)
         |        Size: 1..32
         +-- -R-- String    yadroSoftwareVersion(2)
         +-- -R-- EnumVal   yadroSoftwarePurpose(3)
         |        Values: unknown(0), other(1), system(2), bmc(3), host(4)
         +-- -R-- EnumVal   yadroSoftwareActivation(4)
         |        Values: notReady(0), invalid(1), ready(2), activating(3), active(4), failed(5)
         +-- -R-- Integer32 yadroSoftwarePriority(5)

```

Signed-off-by: Alexander Filippov <a.filippov@yadro.com>

nest1ing committed:
Move YADRO_OID macros to separate file

Signed-off-by: Alexander Filippov <a.filippov@yadro.com>

nest1ing committed:
subagent: remove previous agent implementation

Signed-off-by: Alexander Filippov <a.filippov@yadro.com>

nest1ing committed:
subagent: add initial version

Subagent have no DBus or MIB objects. Only connect to master.

Signed-off-by: Alexander Filippov <a.filippov@yadro.com>

nest1ing committed:
subagent: add yadroHostPowerState field

Add template for export DBus property to SNMP scalar

Signed-off-by: Alexander Filippov <a.filippov@yadro.com>

nest1ing committed:
subagent: add sensors tables

Add C++ templates for export DBus tree as snmp table

Signed-off-by: Alexander Filippov <a.filippov@yadro.com>

nest1ing committed:
subagent: add software table

Signed-off-by: Alexander Filippov <a.filippov@yadro.com>

AlexanderAmelkin commented on agent/yadro/software.cpp:
```diff
@@ -0,0 +1,203 @@
+                      {"NotReady", 0},
+                      {"Invalid", 1},
+                      {"Ready", 2},
+                      {"Acticating", 3},
```
Acti**c**ating? Did you mean `Activating` ?

AlexanderAmelkin commented on agent/yadro/software.cpp:
```diff
@@ -0,0 +1,203 @@
+    softwareTable("/xyz/openbmc_project/software", "",
+                  {
+                      "xyz.openbmc_project.Software.Version",
+                      "xyz.openbmc_project.Software.Actication",
```
What's wrong with this word? ;)

AlexanderAmelkin commented on agent/yadro/software.cpp:
```diff
@@ -0,0 +1,203 @@
+                                          uint8_t>(path, // Object path
+                                                   "",   // No subfolder
+                                                   0xFF, // Purpose
+                                                   0xFF, // Actication
```
Actication again?

AlexanderAmelkin commented on configure.ac:
```diff
@@ -1,28 +1,26 @@
-AC_INIT([phosphor-snmp], [1.0], [https://github.com/YADRO-KNS/phosphor-snmp])
+AC_INIT([phosphor-snmp], 
+        [m4_esyscmd_s([git describe --dirty --long --always])], 
+        [https://github.com/YADRO-KNS/phosphor-snmp/issue])
```
Why did you change the URL to `/issue`? It is invalid, by the way. It should have been `/issues`, but I don't understand the reason anyway.

nest1ing replied:
In this field should be specified `bug-report url`.

AlexanderAmelkin commented on yadro/watcher.cpp:
```diff
@@ -116,68 +119,11 @@ void dbuswatcher::updateSensors(void)
-        "xyz.openbmc_project.Inventory.Manager",
-        "/xyz/openbmc_project/inventory/system/chassis/motherboard/opfw",
-        "xyz.openbmc_project.Inventory.Decorator.Revision", "Version"));
+    // TODO
```
Ehm... we did have BMC version, now we don't have anything?

nest1ing replied:
Now software with version and type present as row in `yadroSoftwareTable`.

nest1ing committed:
subagent: make packages optional

Removes configuration for `snmpd`.
Adds configure switches for disable `agent` or `cfg-manager`.
This functionality may be used by `yocto`.

Signed-off-by: Alexander Filippov <a.filippov@yadro.com>

artemsen commented on agent/yadro/sensors.cpp:
```diff
@@ -0,0 +1,392 @@
+                break;
+            case COLUMN_YADROSENSOR_CRITHIGH:
+                snmp_set_var_typed_integer(request->requestvb, ASN_INTEGER,
+                                           std::get<7>(data));
```
Magic numbers for std::get<>(). May be we can use some identifiers (macro?) for them?

AlexanderAmelkin reviewed (dismissed):
I'm fine with this change, but I'd like @artemsen to approve as well.

AlexanderAmelkin reviewed (commented):
>  subagent: makes packages optional

Make it imperative, not 3rd person. That is, "Make packages optional"

> This functional may be used by `yocto`.

It's either "function" or "functionality", not "functional" (which is noun for a math function, or is an adjective).

artemsen commented on agent/data/table.hpp:
```diff
@@ -0,0 +1,300 @@
+    }
+
+    /**
+     * Register MIB handlers
```
Function parameters description is missing.

artemsen commented on agent/main.cpp:
```diff
@@ -0,0 +1,151 @@
+{
+    if (parse_args(argc, argv) != 0)
+    {
+        print_usage();
```
It's a bad idea to print usage help on any errors.
By the code, function snmp_log_options() can return error but printing help is ambiguous in this case.

artemsen commented on agent/main.cpp:
```diff
@@ -0,0 +1,151 @@
+    do
+    {
+        snmpagent_run();
+    } while ((rc = evt.run(-1)) >= 0);
```
Bad style, I would move `rc = evt.run(-1)` to the cycle body.

artemsen commented on agent/yadro/sensors.cpp:
```diff
@@ -0,0 +1,392 @@
+        switch (type[0])
+        {
+            case 't': // temperature
+                snprintf(fieldOidStr, sizeof(fieldOidStr) - 1,
```
It would be better to make a function instead of copy/paste.

You have agent/yadro/yadro_oid.hpp with YADRO_OID macro:
#define YADRO_OID(args...)                                                     \
+    {                                                                          \
+        1, 3, 6, 1, 4, 1, 49769, ##args                                        \
+    }

artemsen commented on agent/yadro/sensors.cpp:
```diff
@@ -0,0 +1,392 @@
+     */
+    state_t getState() const
+    {
+        if (std::get<8>(data))
```
Why not 42? ;) Magic numbers are ugly.

artemsen commented on agent/yadro/sensors.cpp:
```diff
@@ -0,0 +1,392 @@
+    {
+        switch (type[0])
+        {
+            case 't':
```
What the magic case?!

artemsen commented on agent/yadro/sensors.cpp:
```diff
@@ -0,0 +1,392 @@
+    DEBUGMSGTL(("yadro:init", "Initialize yadroSensors\n"));
+    update();
+
+    tempSensors.init_mib(
```
Something wrong. Why didn't you use a factory or an object manager to work with different sensor types through one entry point?

artemsen commented on agent/yadro/software.cpp:
```diff
@@ -0,0 +1,203 @@
+        phosphor::snmp::data::table::Item<std::string, uint8_t, uint8_t,
+                                          uint8_t>(path, // Object path
+                                                   "",   // No subfolder
+                                                   0xFF, // Purpose
```
Magic number: 0xFF. Anyway, why 0xff? Why isn't default 0?

artemsen commented on sdbusplus/helper.hpp:
```diff
@@ -0,0 +1,146 @@
+    using Interfaces = std::vector<Interface>;
+
+    /** @brief Get subtree from the mapper. */
+    static auto getSubTree(const std::string& path, const Interfaces& ifaces,
```
I think you can replace all input parameters to auto, too ;)
`auto` in function definition is ambiguous, it's hard to read such code.

artemsen commented on sdevent/event.hpp:
```diff
@@ -0,0 +1,183 @@
+     *         Zero if timeout reached and positive value if an event was
+     *         dispatched.
+     */
+    int run(uint64_t usec = -1)
```
I know that systemd interface use such method with negative unsigned number. But for our code I prefer to use macro, something like `#define WAIT_INFINITE -1`

artemsen commented on agent/yadro/sensors.cpp:
```diff
@@ -0,0 +1,392 @@
+    {
+        switch (type[0])
+        {
+            case 't': // temperature
```
Use defines for types instead of comment.

artemsen commented on agent/yadro/sensors.cpp:
```diff
@@ -0,0 +1,392 @@
+            switch (type[0])
+            {
+                case 't': // temperature
+                    notifyOid.assign(YADRO_OID(0, 2));
```
Specify notifyOid as class member.

artemsen commented on agent/yadro/sensors.cpp:
```diff
@@ -0,0 +1,392 @@
+        {
+            switch (type[0])
+            {
+                case 't': // temperature
```
Magic 't'

artemsen commented on agent/main.cpp:
```diff
@@ -0,0 +1,161 @@
+                break;
+
+            case 'h':
+                rc = 1;
```
Magic constant

artemsen commented on agent/snmp.cpp:
```diff
@@ -0,0 +1,113 @@
+static void sdevent_snmp_update()
+{
+    int maxfd = 0;
+    int block = 1;
```
If it is a bool use name like is_blocked.

artemsen commented on agent/yadro/sensors.cpp:
```diff
@@ -0,0 +1,392 @@
+            {
+                size_t stateOidLen = MAX_OID_LEN;
+                stateOid.resize(stateOidLen);
+                char stateOidStr[4096] = {0};
```
Decrease size

artemsen commented on agent/yadro/sensors.cpp:
```diff
@@ -0,0 +1,392 @@
+    template <size_t Idx> int getValue() const
+    {
+        return static_cast<int>((std::get<Idx>(data) + 0.5) /
+                                std::get<9>(data));
```
9?

artemsen commented on agent/yadro/sensors.cpp:
```diff
@@ -0,0 +1,392 @@
+     */
+    template <size_t Idx> int getValue() const
+    {
+        return static_cast<int>((std::get<Idx>(data) + 0.5) /
```
Use round()

nest1ing commented:
@AlexanderAmelkin - your turn

nest1ing commented:
I found and fix warning in `smilint YADRO-MIB.txt`
